src/Acl.php

```diff
@@ -45,9 +60,30 @@ public function getRules(Model $model)
                 $modelClasses[] = $class;
             }
         } while (($class = get_parent_class($class)) !== false);
-        $res = $user->ref('AccessRules')->addCondition('model', 'in', $modelClasses);
 
-        return $res; // @phpstan-ignore-line
+        // Internally disable ACL for a moment and limit to only required fields to avoid recursion
+        $this->disabled = true;
+        $rules = $user->ref('AccessRules')
+            ->addCondition('model', 'in', $modelClasses)
+            ->export(['model', 'all_visible', 'visible_fields', 'all_editable', 'editable_fields', 'all_actions', 'actions', 'conditions']);
+
+        // normalize
+        foreach ($rules as $k => $rule) {
+            $rules[$k]['visible_fields'] = $rule['all_visible'] ? [] : $this->normalizeValue($rule, 'visible_fields');
+            $rules[$k]['editable_fields'] = $rule['all_editable'] ? [] : $this->normalizeValue($rule, 'editable_fields');
+            $rules[$k]['actions'] = $rule['all_actions'] ? [] : $this->normalizeValue($rule, 'actions');
+        }
+
+        $this->disabled = false;
+
+        return $rules;
+    }
+
+    private function normalizeValue(array $rule, string $field): array
+    {
+        $v = $rule[$field] ?? [];
+
+        return is_array($v) ? $v : explode(',', $v);
```

Comment:
```suggestion
        return is_array($v) ? ($v ?? []) : explode(',', $v);
```

the arg should be `mixed $value` and no field